### ISSUE TITLE
Override MemoryStream.ToString

### ISSUE THREAD

**Issue by GSPP** [dotnet/runtime#64512]
In order to make debugging easier, `MemoryStream` should have a useful `ToString` output or a debugger display. My suggestion:

> MemoryStream(Length = 42, Position = 0, Writable = true, Capacity = 64, Data = 0x123456...)

Design considerations:

1. Is compatibility a problem? If no, that would speak for `ToString`. If yes, then `DebuggerDisplay`.
2. Does this cause problems with derived classes?
2. I added `Writable` because that can help catch bugs.
3. Display the first N bytes of data, after that render the truncation suffix `...`. This can help debug problems when the data is obviously incorrect (such as all zeroes, where real data was expected). Maybe N = 8?
4. Data could be rendered as `$"0x{Hex} ('{ASCII}')"`. Adding a textual representation could help identify the data, for example GIF files start with `'GIF'`.
5. When disposed, revert to the default display. I don't have a better idea of what to do here.

**Comment by Jozkee** [comment]
I would personally prefer a DebuggerDisplay before a ToString override but I think both options are feasible. cc @bartonjs.
Linking related Framework Design Guidelines: https://stackoverflow.com/a/10278430/4231460.

**Comment by bartonjs** [comment]
I think a DebuggerDisplay is more appropriate than a ToString.

I don't think the actual contents are interesting to display, as that feels very wasteful; and can be better accomplished by just looking at stream._buffer in a memory inspection window.

**Comment by stephentoub** [comment]
I agree with @bartonjs.  This does not warrant overriding ToString.  If we want to add something here, it should be a [DebuggerDisplay], though for me personally even that feels like it adds minimal value, as everything suggested to be included is visible with one or two clicks:
<img width="545" alt="image" src="https://user-images.githubusercontent.com/2642209/157745429-bc9f22eb-146a-4b27-b11f-2cd02ecfdc0e.png">

**Comment by GSPP** [comment]
If the data is not ASCII decodable, it should receive some (basically arbitrary) default treatment. Some use cases that I thought of:

1. JSON starts with `{`
2. HTML starts with `<html>`
3. GIF is `GIF89`
4. Many other magic file headers
5. Compressed or encrypted data appears as gibberish

All of this is meaningful to debug real-world problems. Looking at the `_buffer` field does not offer a human-understandable view. It lists a series of bytes formatted as decimal. Thinking about this... This should be changed, too. The debugger should show something interpretable for binary data.

I believe, that ASCII encoding in the framework supports this `?` decoding built-in, so it would not require much effort to implement.

**Comment by bartonjs** [comment]
For the current experience: Open a memory watch window:

![image](https://user-images.githubusercontent.com/10642668/158494210-f850e45b-9f59-4e5b-829c-ae73f3b224ed.png)

In the address box, give it the symbolic name (e.g. `memoryStream._buffer`):

![image](https://user-images.githubusercontent.com/10642668/158494338-d35cc6e7-aa87-4586-9b0d-1a2a279e9e0b.png)

Hit enter:

![image](https://user-images.githubusercontent.com/10642668/158494491-61faa589-f611-41eb-8df3-a146288cf3e7.png)

(Looks like that's showing me more of the object header than I'd have expected, but it's not too far off from where I'd've expected it to be)

Anything fancy around content sniffing and ASCII decoding (already exists) and such feels like debugger features more than runtime features.  And why would we do anything for MemoryStream specifically? It's just as relevant to `byte[]`/`Span<byte>`/`ReadOnlySpan<byte>`... which is why it feels more like a debugger feature to me.

Promoting some things like Length/Position into a DebuggerDisplay I (personally) much more readily accept.  I've probably watched the Position value of a seekable stream over the contents of any buffer it is holding at about 1000:1.  For the few times I've cared about the contents, there's the memory view window.

**Comment by adamsitnik** [comment]
> Anything fancy around content sniffing and ASCII decoding (already exists) and such feels like debugger features more than runtime features.

I totally agree with @bartonjs.

We could add `[DebuggerDisplay]` to `MemoryStream`. But why just to `MemoryStream`? What about other `Stream` derived types?

But then we could use very few properties (`CanRead`, `CanWrite` ?). Other like `Stream.CanSeek` can perform a sys-call:

https://github.com/dotnet/runtime/blob/631389fb941eb618fef883975ca5f4f5b03a93cf/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/OSFileStreamStrategy.cs#L71

https://github.com/dotnet/runtime/blob/631389fb941eb618fef883975ca5f4f5b03a93cf/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs#L28

and potentially throw.

I am not convinced that we should do it.
